src/sentry/models/identity.py

```diff
@@ -59,24 +61,56 @@ def get_provider(self):
 
 
 class IdentityManager(BaseManager):
-    def reattach(
+    def get_identities_for_user(self, user: User, provider: ExternalProviders) -> QuerySet:
+        return self.filter(user_id=user.id, idp__type=provider.name)
+
+    def has_identity(self, user: User, provider: ExternalProviders) -> bool:
+        return self.get_identities_for_user(user, provider).exists()
+
+    def link_identity(
         self,
-        idp: "IdentityProvider",
+        user: User,
+        idp: IdentityProvider,
         external_id: str,
-        user: "User",
-        defaults: Mapping[str, Any],
-    ) -> "Identity":
+        should_reattach: bool = True,
+        defaults: Mapping[str, Any | None] = None,
+    ) -> Identity:
         """
-        Removes identities under `idp` associated with either `external_id` or `user`
-        and creates a new identity linking them.
+        Link the user with the identity. If `should_reattach` is passed, handle
+        the case where the user is linked to a different identity or the
+        identity is linked to a different user.
         """
-        lookup = Q(external_id=external_id) | Q(user=user)
-        self.filter(lookup, idp=idp).delete()
+        defaults = {
+            **(defaults or {}),
+            "status": IdentityStatus.VALID,
+            "date_verified": timezone.now(),
+        }
+        try:
+            identity, created = self.get_or_create(
+                idp=idp, user=user, external_id=external_id, defaults=defaults
+            )
+            if not created:
+                identity.update(**defaults)
+        except IntegrityError as e:
```

Comment:
When would there be an integrity error?

Reply:
I'm not sure, I copied it from the usages in the link_identity files.